modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,1138 @@
+ *      \f[q = w + x\boldsymbol{i} + y\boldsymbol{j} + z\boldsymbol{k}\f]
+ *      \f[q = [w, x, y, z]\f]
+ *      \f[q = [w, \boldsymbol{v}] \f]
+ *      \f[q = ||q||[\cos\theta, u_x\sin\theta,u_y\sin\theta,  u_z\sin\theta].\f]
+ *      \f[q = ||q||[\cos\theta, \boldsymbol{u}\sin\theta]\f]
+ * where \f$\theta = \frac{\psi}{2}\f$, \f$\psi\f$ represents rotation angle,
+ * \f$\boldsymbol{u} = [u_x, u_y, u_z]\f$ represents normalized rotation axis,
+ * and \f$||q||\f$ represents the norm of \f$q\f$
+ *
+ * A unit quaternion is usually represents rotation, which has the form:
+ *      \f[q = [\cos\theta, u_x\sin\theta,u_y\sin\theta,  u_z\sin\theta].\f]
+ *
+ * To create a quaternion representing the rotation around the axis \f$\boldsymbol{v}\f$
+ * with angle \f$\psi\f$, you can use
+ * ```
+ * using namespace cv;
+ * double angle = CV_PI;
+ * Vec3d axis = {0, 0, 1};
+ * Quatd q = Quatd::createFromAngleAxis(angle, axis);
+ * ```
+ *
+ * You can simply use four same type number to create a quaternion
+ * ```
+ * Quatd q(1, 2, 3, 4);
+ * ```
+ * Or use a Vec4d or Vec4f vector.
+ * ```
+ * Vec4d vec{1, 2, 3, 4};
+ * Quatd q(vec);
+ * ```
+ *
+ * ```
+ * Vec4f vec{1, 2, 3, 4};
+ * Quatf q(vec);
+ * ```
+ *
+ * If you already have a 3x3 rotation matrix R, then you can use
+ * ```
+ * Quatd q = Quatd::createFromRotMat(R);
+ * ```
+ *
+ * If you already have a rotation vector rvec which has the form of `angle * axis`, then you can use
+ * ```
+ * Quatd q = Quatd::createFromRvec(rvec);
+ * ```
+ *
+ * To extract the rotation matrix from quaternion, see toRotMat3x3()
+ *
+ * To extract the Vec4d or Vec4f, see toVec()
+ *
+ * To extract the rotation vector, see toRotVec()
+ *
+ * If there are two quaternions \f$q_0, q_1\f$ are needed to interpolate, you can use nlerp(), slerp() or spline()
+ * ```
+ * Quatd::nlerp(q0, q1, t)
+ *
+ * Quatd::slerp(q0, q1, t)
+ *
+ * Quatd::spline(q0, q0, q1, q1, t)
+ * ```
+ * spline can smoothly connect show the rotation between multiple quaternions
+ *
+ * Three ways to get a element in Quaternion
+ * ```
+ * Quatf q(1,2,3,4);
+ * std::cout << q.w << std::endl; // w=1, x=2, y=3, z=4
+ * std::cout << q[0] << std::endl; // q[0]=1, q[1]=2, q[2]=3, q[3]=4
+ * std::cout << q.at(0) << std::endl;
+ * ```
+ */
+template <typename _Tp> class Quat;
+template <typename S>
+std::ostream& operator<<(std::ostream&, const Quat<S>&);
+
+template <typename _Tp>
+class Quat
+{
+static_assert(std::is_floating_point<_Tp>::value, "Quaternion only make sense with type of float and double");
+using value_type = _Tp;
+public:
+    Quat() = default;
+
+    /**
+     * @brief From Vec4d or Vec4f.
+     */
+    explicit Quat(const Vec<_Tp, 4> &coeff);
+
+    /**
+     * @brief from four numbers.
+     */
+    Quat(_Tp w, _Tp x, _Tp y, _Tp z);
+
+    /**
+     * @brief from an angle, axis. Axis will be normalized in this function. And
+     * it generates
+     * \f[q = [\cos\theta, u_x\sin\theta,u_y\sin\theta,  u_z\sin\theta].\f]
```

Comment:
Please fix: `theta` => `theta/2` as it should be

Reply:
OK.  I will improve the definition of variables in each function.